gbench/syclblas/main.cpp

```diff
@@ -2,68 +2,80 @@
 #include "utils.hpp"
 
 namespace Private {
+// Create a shared pointer to a sycl blas executor, so that we don't keep
+// reconstructing it each time (which is slow). Although this won't be
+// cleaned up if RunSpecifiedBenchmarks exits badly, that's okay, as those
+// are presumably exceptional circumstances.
 ExecutorPtr ex;
-// Declare the queue globally so that it is properly destroyed if exit() is
-// called. If not, then it will not be destroyed before the sycl runtime tries
-// to exit, which will cause an exception.
-cl::sycl::queue q;
+std::unique_ptr<cli_device_selector> cdsp;
+void free_device_selector() { cdsp.reset(); }
 }  // namespace Private
 
 ExecutorPtr getExecutor() { return Private::ex; }
 
 int main(int argc, char** argv) {
-  {
-    // Initialise the command line device selector and queue in a nested scope
-    // so that the cds is destroyed before benchmark::Initialize is called. If
-    // not, then if benchmark::Initialize calls exit(), the cds will not be
-    // freed before the sycl runtime tries to exit, leading to an exception, as
-    // the cds will still hold some sycl objects.
-    cli_device_selector cds(argc, argv);
-    Private::q =
-        cl::sycl::queue(cds, {cl::sycl::property::queue::enable_profiling()});
-  }
+  // Initialise the command line device selector in a unique pointer so that we
+  // can register an `atexit` handler to delete the command line device
+  // selector. If not, then if benchmark::Initialize calls exit() (for example,
+  // if the flag `--help` is passed), the cds will not be freed before the sycl
+  // runtime tries to exit, leading to an exception, as the cds will still hold
+  // some sycl objects.
+  Private::cdsp =
+      std::unique_ptr<cli_device_selector>(new cli_device_selector(argc, argv));
+  std::atexit(Private::free_device_selector);
 
   // Initialise googlebench
   benchmark::Initialize(&argc, argv);
 
+  // Create a queue from the device selector - do this after initialising
+  // googlebench, as otherwise we may not be able to delete the queue before we
+  // exit (if Initialise calls exit(0))
+  cl::sycl::queue q = cl::sycl::queue(
+      *Private::cdsp.get(), {cl::sycl::property::queue::enable_profiling()});
+
 // Print out some information about the device that we're running on.
 #if 1
```

Comment:
Good question. I suspect so that the debugging/device dumping code can easily be removed without commenting it out. I think it might be best to just leave the code in by default, so I'll remove the preprocessor directives.